src/main/java/ru/spbstu/recognizer/GibbsMotifRecognizer.java

```diff
@@ -11,49 +11,102 @@
+        int nextRand = random.nextInt(records.size() - 1);
+        FastaRecord nextRecord = records.get(nextRand);
 
-        while ((delta > desiredDelta) || Double.isInfinite(delta)) {
-            // to not repeat random number in loop
-            while (nextRand == prevRand) {
-                nextRand = random.nextInt(records.size() - 1);
-            }
-            prevRand = nextRand;
+        // find Motif with the highest score
+        DatasetScore resultScore = smCalculator.calculateScoreForRecord(nextRecord, matrix);
+        ScoredMotif maxScoredMotif = resultScore.getMax();
 
-            // random record
-            FastaRecord nextRecord = records.get(nextRand);
+        // Change found Motif with one with same recordID
+        motifSet.setMotif(maxScoredMotif);
 
-            // find Motif with the highest score
-            DatasetScore resultScore = smCalculator.calculateScoreForRecord(nextRecord, matrix);
-            ScoredMotif maxScoredMotif = resultScore.getMax();
+        // update delta
+        PWMatrix newMatrix = pwCalculator.calculateMatrix(motifSet);
+        double delta = countDelta(matrix, newMatrix);
+        matrix = newMatrix;
 
-            // Change found Motif with one with same recordID
-            motifSet.setMotif(maxScoredMotif);
+        return delta;
+    }
 
-            // update delta
-            PWMatrix newMatrix = pwCalculator.calculateMatrix(motifSet);
-            delta = matrix.countDelta(newMatrix);
-            matrix = newMatrix;
+    public double countDelta(PWMatrix first, PWMatrix second) throws InvalidAlgorithmParameterException {
+        if (first.getSequenceLength() != second.getSequenceLength()) {
+            throw new InvalidAlgorithmParameterException();
         }
 
-        // recount matrix cause relative format is more human-readable
-        pwCalculator = new PWCalculator(new CalculationStrategy
-                .Builder(Constants.DEFAULT_WINDOW_SIZE).calculationType(Constants.calculationMethod.RELATIVE).build());
-        return pwCalculator.calculateMatrix(motifSet);
+        double delta = 0.0;
+
+        for (char nucle: Constants.NUCLEOTIDES) {
+            double[] firstRow = first.getRowByNucleotide(nucle).getRow();
+            double[] secondRow = second.getRowByNucleotide(nucle).getRow();
+
+            for (int i = 0; i < firstRow.length; i ++) {
+                if (!((Double.isInfinite(firstRow[i]) && Double.isInfinite(secondRow[i])))) {
+                    delta += Math.abs((firstRow[i] - secondRow[i]));
+                }
+            }
+        }
+
+        return delta;
+    }
+
+
+    private class Randomizer {
+
+        private int prevInt = -1;
+        private Random random = new Random();
+
+        public int nextInt(int max) {
+            int nextRand = prevInt;
+
+            // to not repeat random number
+            while (nextRand == prevInt) {
+                nextRand = random.nextInt(max);
+            }
```

Comment:
prevInt contains only the previous value. So the random values can repeat. In order to have unique values we need to keep all used values but I don't think we really need it.